File: src/APIClient.ts
import { getBackendSrv } from '@grafana/runtime';
import { FieldType, MutableDataFrame } from '@grafana/data';
import cache from 'memory-cache';
import { Pair, QueryResponse } from 'types';
let _lodash = require('lodash');
import _ from 'lodash';
import { lastValueFrom } from 'rxjs';

let _lodash2 = _interopRequireDefault(_lodash);
function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}
import * as utils from './Utils';
export class APIClient {
  requestOptions: { headers: any; withCredentials: boolean; url: string };
  cache: any;
  lastCacheDuration: number | undefined;
  cacheTimeout: number;
  constructor(headers: any, withCredentials: boolean, url: string, cacheTimeout: number) {
    this.requestOptions = {
      headers: headers,
      withCredentials: withCredentials,
      url: url,
    };
    this.cacheTimeout = cacheTimeout;
    this.cache = new cache.Cache();
  }
  async cachedGet(
    method: string,
    path: string,
    params: Array<Pair<string, string>>,
    cacheDurationSeconds: number | null,
    headers?: Array<Pair<string, string>>,
    body?: string,
    options?: any
  ) {
    let cacheTime = 60;
    if (typeof cacheDurationSeconds === 'undefined' || !cacheDurationSeconds) {
      cacheTime = this.cacheTimeout;
    } else {
      cacheTime = cacheDurationSeconds;
    }
    console.log('using cache timeout: ', cacheTime);

    console.log('new this.requestOptions.url: ', this.requestOptions.url);
    console.log('new path: ', path);
    let cacheKey = this.requestOptions.url + path;

    cacheKey += '/body/' + body;
    let cacheKeyNoTime = cacheKey;
    if (params && Object.keys(params).length > 0) {
      cacheKey =
        cacheKey +
        (cacheKey.search(/\?/) >= 0 ? '&' : '?') +
        params.map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`).join('&');
    }

    if (this.lastCacheDuration !== cacheTime) {
      this.cache.del(cacheKey);
    }
    this.lastCacheDuration = cacheTime;

    let cachedItem = this.cache.get(cacheKey);

    if (!cachedItem && cacheKey.includes('?')) {
      let cacheKeys = this.cache.keys();
      for (let i = 0; i < cacheKeys.length; i++) {
        let key = cacheKeys[i];
        if (key.includes(cacheKeyNoTime) && key.includes('?')) {
          let cacheTimeParams: any = key.substring(key.indexOf('?') + 1, key.length);
          cacheTimeParams = cacheTimeParams.split('&');
          let cacheStartTime = cacheTimeParams[0].substring(
            cacheTimeParams[0].indexOf('=') + 1,
            cacheTimeParams[0].length
          );
          let cacheEndTime = cacheTimeParams[1].substring(
            cacheTimeParams[1].indexOf('=') + 1,
            cacheTimeParams[1].length
          );
          let timeParams: any = cacheKey.substring(cacheKey.indexOf('?') + 1, cacheKey.length);
          timeParams = timeParams.split('&');
          let startTime = timeParams[0].substring(timeParams[0].indexOf('=') + 1, timeParams[0].length);
          let endTime = timeParams[1].substring(timeParams[1].indexOf('=') + 1, timeParams[1].length);
          let startTimeDifference = startTime - cacheStartTime;
          let endTimeDifference = endTime - cacheEndTime;
          if (startTimeDifference >= 0) {
            if (startTimeDifference <= cacheTime * 1000 && endTimeDifference <= cacheTime * 1000) {
              console.log('cache item found in timerange');
              cachedItem = this.cache.get(key);
              break;
            }
          }
        }
      }
    }

    if (cachedItem) {
      console.log('cache item found');
      return Promise.resolve(cachedItem);
    }

    let paramString: any = '';
    if (params.length > 0) {
      paramString = '?' + params.map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`).join('&');
    }

    let result: any = '';
    if (method === 'GET') {
      result = getBackendSrv().get(this.requestOptions.url + path, paramString);
    } else if (method === 'POST') {
      result = getBackendSrv().post(this.requestOptions.url + path + paramString, body);
    }

    // Deprecated method
    //const result = getBackendSrv().datasourceRequest(options);

    this.cache.put(cacheKey, result, cacheTime * 1000);

    return result;
  }
  request(options) {
    // This is the NEW request method
    options.withCredentials = this.requestOptions.withCredentials;
    options.headers = this.requestOptions.headers;
    let apiPath = options.url;
    options.url = this.requestOptions.url + apiPath;
    const response = getBackendSrv().fetch<any>({
      ...options,
    });
    return lastValueFrom(response);
    // OLD CACHE METHOD
    //
    // let paramStartIndex = apiPath.indexOf('?');
    // if (paramStartIndex === -1) {
    //   paramStartIndex = apiPath.length;
    // }
    // let path = apiPath.substring(0, paramStartIndex);
    // let paramsObject: Array<Pair<string, string>> = [];
    // if (options.url.indexOf('?') !== -1) {
    //   let paramStr = options.url.substring(options.url.indexOf('?') + 1, options.url.length);
    //   let paramArray = paramStr.split('&');
    //   paramArray.map((value) => {
    //     let key = value.substring(0, value.indexOf('='));
    //     let keyValue = value.substring(value.indexOf('=') + 1, value.length);
    //     let pair: Pair<string, string> = [key, keyValue];
    //     paramsObject.push(pair);
    //   });
    // }
    // if (options.cacheOverride) {
    //   let cacheSecondIndex = options.cacheOverride.indexOf('s');
    //   let cacheMinuteIndex = options.cacheOverride.indexOf('m');
    //   if (cacheSecondIndex !== -1) {
    //     options.cacheOverride = parseInt(options.cacheOverride.substring(0, cacheSecondIndex), 10);
    //   } else if (cacheMinuteIndex !== -1) {
    //     options.cacheOverride = parseInt(options.cacheOverride.substring(0, cacheMinuteIndex), 10) * 60;
    //   }
    // }
    // return this.cachedGet(
    //   options.method,
    //   path,
    //   paramsObject,
    //   options.cacheOverride,
    //   options.headers,
    //   options.data,
    //   options
    // );
  }
  // mapAlertTags(response) {
  //   let tags: any = [];
  //   response.map((d) => {
  //     if (typeof d.additional_info === 'undefined') {
  //       return;
  //     }
  //     try {
  //       let additional_info = JSON.parse(d.additional_info);
  //       let keys = Object.keys(additional_info);
  //       let tagKeys = keys.filter((k) => {
  //         return k.includes('tbac-');
  //       });
  //       tagKeys.map((k) => {
  //         tags.push({ key: k, value: additional_info[k] });
  //       });
  //     } catch (e) {
  //       console.log(e);
  //     }
  //   });
  //   tags = tags.filter(
  //     (option, index, self) => index === self.findIndex((t) => t.value === option.value && t.key === option.key)
  //   );
  //   return tags;
  // }
  mapResponseToVariable(result) {
    let resultsParsed = _lodash2.default.map(result, function (d, i) {
      if (typeof d.name !== 'undefined' && typeof d.id !== 'undefined') {
        if (d.name === '' || d.name === null) {
          d.name = 'NULL';
        }
        if (d.id === '' || d.id === null) {
          d.id = 'NULL';
        }
        return { text: d.name, value: d.id };
      } else {
        let keys = Object.keys(d);
        if (d[keys[0]] === '' || d[keys[0]] === null) {
          d[keys[0]] = 'NULL';
        }
        if (keys[1] && (d[keys[1]] === '' || d[keys[1]] === null)) {
          d[keys[1]] = 'NULL';
        }
        return { text: d[keys[0]], value: keys[1] ? d[keys[1]] : d[keys[0]] };
      }
    });

    const hasNull = resultsParsed.some(item => item.text === 'NULL' && item.value === 'NULL');

    if (!hasNull) {
      resultsParsed.push({ text: 'NULL', value: 'NULL' });
    }

    resultsParsed.push({ text: '*', value: '*' });
    return resultsParsed;
  }
  mapChecksToValue(result) {
    return _lodash2.default.map(result, function (d, i) {
      if (typeof d.name !== 'undefined' && typeof d.id !== 'undefined') {
        if (d.name === '' || d.name === null) {
          d.name = 'NULL';
        }
        if (d.id === '' || d.id === null) {
          d.id = 'NULL';
        }
        return { label: d.name, value: d.id };
      } else {
        let keys = Object.keys(d);
        if (d[keys[0]] === '' || d[keys[0]] === null) {
          d[keys[0]] = 'NULL';
        }
        if (keys[1] && (d[keys[1]] === '' || d[keys[1]] === null)) {
          d[keys[1]] = 'NULL';
        }
        return { label: d[keys[0]], value: keys[1] ? d[keys[1]] : d[keys[0]] };
      }
    });
  }
  mapChecksToValuePlusSuffix(result) {
    return _lodash2.default.map(result, function (d, i) {
      let keys = Object.keys(d);
      return { label: d[keys[0]], value: keys[1] ? d[keys[1]] : d[keys[0]], suffix: d[keys[2]] };
    });
  }
  mapValueSuffixToColumns(result) {
    let displayArray = _lodash2.default.map(result, (d, i) => {
      return { label: d.label + ':display', value: d.value + ':d' };
    });
    let valueArray = _lodash2.default.map(result, (d, i) => {
      return { label: d.label + ':value', value: d.value + ':v' };
    });
    let finalResult = displayArray.concat(valueArray);
    finalResult = _.orderBy(finalResult, ['label'], ['asc']);
    return finalResult;
  }
  mapValueAsSuffix(result, addType) {
    let options = _lodash2.default.map(result, (d) => {
      let option: any = {
        label: addType ? d.label + ' (' + d.type + ')' : d.label,
        value: d.value,
        description: d.value,
      };
      if (typeof d.options !== 'undefined') {
        option.options = _lodash2.default.map(d.options, (n) => {
          return { label: addType ? n.label + ' (' + n.type + ')' : n.label, value: n.value, description: n.value };
        });
      }
      return option;
    });
    options = _.orderBy(options, ['label'], ['asc']);
    return options;
  }
  mapSuffixToLabel(result) {
    return _lodash2.default.map(result, (d) => {
      return { label: d.label + ' (' + d.suffix + ')', value: d.value };
    });
  }
  appendInstanceNameToResponse(response, instanceName) {
    response = _lodash2.default.map(response, function (d, i) {
      d.instanceName = instanceName;
      return d;
    });
    return response;
  }
  // mapTagsToValue(result) {
  //   let tagsList: any[] = [];
  //   for (let d = 0; d < result.data.length; d++) {
  //     for (let v = 0; v < result.data[d].values.length; v++) {
  //       let tagValue = result.data[d].key.name + ' - ' + result.data[d].values[v].value;
  //       let tagId = result.data[d].values[v].id;
  //       tagsList.push({ text: tagValue, value: tagId });
  //     }
  //   }
  //   return tagsList;
  // }
  mapToTextValue(result) {
    return _lodash2.default.map(result, function (d, i) {
      if (d && d.text && d.value) {
        return { text: d.text, value: d.value };
      } else if (_lodash2.default.isObject(d)) {
        return { text: d, value: i };
      }
      return { text: d, value: d };
    });
  }
  mapOutageResponseToFrame(result, target) {
    return result.map((data) => {
      let ciName = data.ci;
      console.log(ciName);
      return utils.parseResponse(data.datapoints, ciName, target, [], FieldType.string);
    });
  }
  mapTrendResponseToFrame(result, target) {
    return Object.keys(result[0]).map((data) => {
      return utils.parseResponse(result[0][data].datapoints, data, target, [], FieldType.number);
    });

    // return result.map((data) => {

    //   return utils.parseResponse(data.datapoints, '', target, [], FieldType.number);
    // });
  }
  mapMetricsResponseToFrame(result, target) {
    return result.map((data) => {
      let seriesName = data.source + ':' + data.metricName;
      if (data.type.length > 0) {
        seriesName += ':' + data.type;
      }
      return utils.parseResponse(data.datapoints, seriesName, target, [], FieldType.number);
    });
  }
  mapAnamMetricsResponseToFrame(result, target) {
    let response = result.map((r) => {
      let ciName = r.ci_name;
      let metricName = r.metric_name;

      return r.data.series.map((series) => {
        let seriesName = ciName + ':' + metricName + ':' + series.type;
        return utils.parseAnomResponse(series.data, seriesName, target, [], FieldType.number);
      });
    });
    // Flattens the array
    response = [].concat.apply([], response);
    return response;
  }
  mapTextResponseToFrame(result, refId) {
    const frame = new MutableDataFrame({
      fields: [],
      refId: refId,
    });
    if (utils.debugLevel() === 1) {
      utils.printDebug('You are Inside mapTextResponseToFrame');
    }
    console.log(result);
    if (!(result.length > 0)) {
      return [];
    }
    // result = result.map((r) => {
    //   if (r.additional_info) {
    //     let additonal_info = JSON.parse(r.additional_info);
    //     let keys = Object.keys(additonal_info);
    //     let tags = keys.filter((k) => {
    //       return k.includes('tbac-');
    //     });
    //     r.tbac_data = {};
    //     for (let j = 0; j < tags.length; j++) {
    //       r.tbac_data[tags[j]] = additonal_info[tags[j]];
    //     }
    //     r.tbac_data = JSON.stringify(r.tbac_data);
    //     return r;
    //   } else {
    //     return r;
    //   }
    // });
    console.log(result);
    let filedNames = Object.keys(result[0]);
    for (let i = 0; i < filedNames.length; i++) {
      let values = result.map((d) => d[filedNames[i]]);
      if (filedNames[i] === 'new' || filedNames[i] === 'value:display') {
        values = this.sanitizeValues(values);
      }
      let fieldType = FieldType.string;
      if (values.length >= 0) {
        fieldType = utils.getFiledType(values[0], filedNames[i]);
      }
      frame.addField({
        name: filedNames[i],
        type: fieldType,
        values: values,
      });
    }
    if (utils.debugLevel() === 1) {
      utils.printDebug(frame);
    }
    return frame;
  }

  createTopologyFrame(result, refId) {
    const data: QueryResponse[] = [
      {
        columns: [
          { text: 'type' },
          { type: 'time', text: 'Time' },
          { text: 'app' },
          { text: 'target_app' },
          { text: 'req_rate' },
          { text: 'error_rate' },
        ],

        rows: result.rows,
        refId: refId || undefined,
        meta: undefined,
      },
    ];
    console.log('topology frame: ', data);
    return data;
  }

  sanitizeValues(values) {
    let sanitizedArray: any[] = [];
    values.map((value) => {
      while (value.indexOf('[code]') !== -1) {
        let strBeforeCode = value.substring(0, value.indexOf('[code]'));
        let strAfterCode = value.substring(value.indexOf('[/code]') + 7, value.length);
        if (value.indexOf('<a') !== -1) {
          let aElement = value.substring(value.indexOf('<a'), value.indexOf('</a>', value.indexOf('<a')));
          let aValue = aElement.substring(aElement.indexOf('>') + 1, aElement.length);
          if (aValue.indexOf('<') !== -1) {
            aValue = aValue.substring(0, aValue.indexOf('<'));
          }
          value = strBeforeCode + aValue + strAfterCode;
        } else {
          value = strBeforeCode + strAfterCode;
        }
      }
      sanitizedArray.push(value);
    });
    return sanitizedArray;
  }
}

File: pkg/APIClient.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"reflect"
	"sort"
	"strings"
	"time"

	"github.com/grafana/grafana-plugin-sdk-go/data"
)

type RequestOptions struct {
	Headers         map[string]string
	WithCredentials bool
	URL             string
}

type APIClient struct {
	RequestOptions RequestOptions
	CacheTimeout   time.Duration
}

type Option struct {
	Label        string   `json:"label"`
	Value        string   `json:"value"`
	Suffix       string   `json:"suffix,omitempty"`
	Type         string   `json:"type,omitempty"`
	Description  string   `json:"description,omitempty"`
	InstanceName string   `json:"instanceName,omitempty"`
	Options      []Option `json:"options,omitempty"`
}

// Constructor function to initialize the APIClient
func Initialize(headers map[string]string, withCredentials bool, url string, cacheTimeout time.Duration) *APIClient {
	return &APIClient{
		RequestOptions: RequestOptions{
			Headers:         headers,
			WithCredentials: withCredentials,
			URL:             url,
		},
		CacheTimeout: cacheTimeout,
	}
}

// Performs an HTTP Request
func (client *APIClient) Request(method string, endpoint string, body interface{}) ([]byte, error) {
	fullURL := client.RequestOptions.URL + endpoint
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest(method, fullURL, bytes.NewBuffer(jsonBody))
	if err != nil {
		return nil, err
	}

	for key, value := range client.RequestOptions.Headers {
		req.Header.Set(key, value)
	}

	clientHTTP := &http.Client{}
	resp, err := clientHTTP.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, errors.New("request failed: " + resp.Status)
	}

	return io.ReadAll(resp.Body)
}

func MapResponseToVariable(result []map[string]interface{}, asterisk bool, showNull bool) []Option {
	var resultsParsed []Option

	for _, d := range result {
		if name, nameOk := d["name"]; nameOk {
			if id, idOk := d["id"]; idOk {
				nameStr := "NULL"
				idStr := "NULL"
				if name != "" && name != nil {
					nameStr = fmt.Sprintf("%v", name)
				}
				if id != "" && id != nil {
					idStr = fmt.Sprintf("%v", id)
				}
				resultsParsed = append(resultsParsed, Option{Label: nameStr, Value: idStr})
			}
		} else {
			keys := reflect.ValueOf(d).MapKeys()
			if len(keys) > 0 {
				firstKey := fmt.Sprintf("%v", keys[0])
				firstVal := fmt.Sprintf("%v", d[firstKey])

				if firstVal == "" || d[firstKey] == nil {
					firstVal = "NULL"
				}

				secondVal := firstVal
				if len(keys) > 1 {
					secondKey := fmt.Sprintf("%v", keys[1])
					secondVal = fmt.Sprintf("%v", d[secondKey])
					if secondVal == "" || d[secondKey] == nil {
						secondVal = "NULL"
					}
				}

				resultsParsed = append(resultsParsed, Option{Label: firstVal, Value: secondVal})
			}
		}
	}

	if showNull {
		resultsParsed = append(resultsParsed, Option{Label: "NULL", Value: "NULL"})
	}
	if asterisk {
		resultsParsed = append(resultsParsed, Option{Label: "*", Value: "*"})
	}

	return resultsParsed
}

func MapChecksToValue(result []map[string]interface{}) []Option {
	var mappedResults []Option

	for _, d := range result {
		if name, ok := d["name"]; ok && d["id"] != nil {
			if name == "" || name == nil {
				name = "NULL"
			}
			if d["id"] == "" || d["id"] == nil {
				d["id"] = "NULL"
			}
			mappedResults = append(mappedResults, Option{
				Label: name.(string),
				Value: d["id"].(string),
			})
		} else {
			keys := make([]string, 0, len(d))
			for k := range d {
				keys = append(keys, k)
			}
			if d[keys[0]] == "" || d[keys[0]] == nil {
				d[keys[0]] = "NULL"
			}
			if len(keys) > 1 && (d[keys[1]] == "" || d[keys[1]] == nil) {
				d[keys[1]] = "NULL"
			}
			label := d[keys[0]].(string)
			value := label
			if len(keys) > 1 {
				value = d[keys[1]].(string)
			}
			mappedResults = append(mappedResults, Option{
				Label: label,
				Value: value,
			})
		}
	}
	return mappedResults
}

func MapChecksToValuePlusSuffix(result []map[string]interface{}) []Option {
	var mappedResults []Option

	for _, d := range result {
		label := fmt.Sprintf("%v", d["name"])
		value := label
		suffix := ""

		if v, ok := d["id"]; ok {
			value = fmt.Sprintf("%v", v)
		}
		if s, ok := d["suffix"]; ok {
			suffix = fmt.Sprintf("%v", s)
		}

		mappedResults = append(mappedResults, Option{
			Label:  label,
			Value:  value,
			Suffix: suffix,
		})
	}
	return mappedResults
}

func MapValueSuffixToColumns(result []Option) []Option {
	var displayArray, valueArray []Option

	for _, d := range result {
		displayArray = append(displayArray, Option{
			Label: d.Label + ":display",
			Value: d.Value + ":d",
		})
		valueArray = append(valueArray, Option{
			Label: d.Label + ":value",
			Value: d.Value + ":v",
		})
	}

	finalResult := append(displayArray, valueArray...)

	sort.Slice(finalResult, func(i, j int) bool {
		return finalResult[i].Label < finalResult[j].Label
	})

	return finalResult
}

func MapValueAsSuffix(result []Option, addType bool) []Option {
	var options []Option

	for _, d := range result {
		option := Option{
			Label:       d.Label,
			Value:       d.Value,
			Description: d.Value,
		}

		if addType && d.Type != "" {
			option.Label = d.Label + " (" + d.Type + ")"
		}

		if len(d.Options) > 0 {
			for _, nested := range d.Options {
				nestedOption := Option{
					Label:       nested.Label,
					Value:       nested.Value,
					Description: nested.Value,
				}
				if addType && nested.Type != "" {
					nestedOption.Label = nested.Label + " (" + nested.Type + ")"
				}
				option.Options = append(option.Options, nestedOption)
			}
		}
		options = append(options, option)
	}

	// Sort the options by Label field in ascending order
	sort.Slice(options, func(i, j int) bool {
		return options[i].Label < options[j].Label
	})

	return options
}

func MapSuffixToLabel(result []Option) []Option {
	var mappedResults []Option

	for _, d := range result {
		labelWithSuffix := d.Label
		if d.Suffix != "" {
			labelWithSuffix = d.Label + " (" + d.Suffix + ")"
		}
		mappedResults = append(mappedResults, Option{
			Label: labelWithSuffix,
			Value: d.Value,
		})
	}

	return mappedResults
}

func AppendInstanceNameToResponse(response []Option, instanceName string) []Option {
	for i := range response {
		response[i].InstanceName = instanceName
	}
	return response
}

func MapToTextValue(result []interface{}) []Option {
	var mappedResults []Option

	for i, d := range result {
		switch item := d.(type) {
		case map[string]interface{}:
			if text, ok := item["text"]; ok {
				if value, ok := item["value"]; ok {
					mappedResults = append(mappedResults, Option{
						Label: fmt.Sprintf("%v", text),
						Value: fmt.Sprintf("%v", value),
					})
					continue
				}
			}
			mappedResults = append(mappedResults, Option{
				Label: fmt.Sprintf("%v", d),
				Value: fmt.Sprintf("%d", i),
			})

		default:
			mappedResults = append(mappedResults, Option{
				Label: fmt.Sprintf("%v", d),
				Value: fmt.Sprintf("%v", d),
			})
		}
	}

	return mappedResults
}

func MapOutageResponseToFrame(result []map[string]interface{}, target string) []*data.Frame {
	frames := make([]*data.Frame, len(result))

	for i, dataPoint := range result {
		// Retrieve the ciName from the data map if present
		ciName, _ := dataPoint["ci"].(string)
		// Retrieve timeseries data (datapoints)
		timeseries, ok := dataPoint["datapoints"].([][]interface{})
		if !ok {
			continue
		}

		frame := ParseResponse(timeseries, ciName, target, data.FieldTypeString)
		frames[i] = frame
	}

	return frames
}

func MapTrendResponseToFrame(result map[string]map[string]interface{}, targetRefID string) []*data.Frame {
	var frames []*data.Frame

	for dataKey, dataValue := range result {
		// Access datapoints within each key and assert its type
		if dataPoints, ok := dataValue["datapoints"].([][]interface{}); ok {
			frame := ParseResponse(dataPoints, dataKey, targetRefID, data.FieldTypeFloat64)
			frames = append(frames, frame)
		}
	}
	return frames
}

func MapMetricsResponseToFrame(result []map[string]interface{}, targetRefID string) []*data.Frame {
	var frames []*data.Frame

	for _, dataEntry := range result {
		seriesName := dataEntry["source"].(string) + ":" + dataEntry["metricName"].(string)
		if metricType, ok := dataEntry["type"].(string); ok && len(metricType) > 0 {
			seriesName += ":" + metricType
		}

		if datapoints, ok := dataEntry["datapoints"].([][]interface{}); ok {
			frame := ParseResponse(datapoints, seriesName, targetRefID, data.FieldTypeFloat64)
			frames = append(frames, frame)
		} else {
			fmt.Println("Warning: Missing or invalid datapoints in data entry")
		}
	}

	return frames
}

func MapAnamMetricsResponseToFrame(result []map[string]interface{}, targetRefID string) []*data.Frame {
	var frames []*data.Frame

	for _, r := range result {
		// Retrieve ci_name and metric_name
		ciName, ciOk := r["ci_name"].(string)
		metricName, metricOk := r["metric_name"].(string)
		if !ciOk || !metricOk {
			fmt.Println("Warning: Missing ci_name or metric_name in result entry")
			continue
		}

		// Process each series in the 'data.series' field
		if seriesData, ok := r["data"].(map[string]interface{}); ok {
			if seriesArray, ok := seriesData["series"].([]interface{}); ok {
				for _, seriesItem := range seriesArray {
					if seriesMap, ok := seriesItem.(map[string]interface{}); ok {
						// Construct the series name
						seriesType, typeOk := seriesMap["type"].(string)
						if !typeOk {
							fmt.Println("Warning: Missing type in series")
							continue
						}
						seriesName := ciName + ":" + metricName + ":" + seriesType

						if seriesPoints, ok := seriesMap["data"].([][]interface{}); ok {
							frame := ParseAnomResponse(seriesPoints, seriesName, targetRefID, data.FieldTypeFloat64)
							frames = append(frames, frame)
						} else {
							fmt.Println("Warning: Missing or invalid data in series")
						}
					}
				}
			}
		}
	}

	return frames
}

func SanitizeValues(values []string) []string {
	var sanitizedArray []string
	for _, value := range values {
		for strings.Contains(value, "[code]") && strings.Contains(value, "[/code]") {
			strBeforeCode := value[:strings.Index(value, "[code]")]
			strAfterCode := value[strings.Index(value, "[/code]")+7:]

			if strings.Contains(strAfterCode, "<a") && strings.Contains(strAfterCode, "</a>") {
				aElementStart := strings.Index(strAfterCode, "<a")
				aElementEnd := strings.Index(strAfterCode, "</a>") + 4
				aElement := strAfterCode[aElementStart:aElementEnd]

				aValueStart := strings.Index(aElement, ">") + 1
				aValueEnd := strings.LastIndex(aElement, "<")
				aValue := aElement[aValueStart:aValueEnd]

				value = strBeforeCode + aValue + strAfterCode[aElementEnd:]
			} else {
				value = strBeforeCode + strAfterCode
			}
		}
		sanitizedArray = append(sanitizedArray, value)
	}
	return sanitizedArray
}

func MapTextResponseToFrame(result []map[string]interface{}, refID string) *data.Frame {
	// Initialize the DataFrame with the reference ID
	frame := data.NewFrame(refID)
	frame.RefID = refID

	// Check if result is empty
	if len(result) == 0 {
		return frame
	}

	// Retrieve the field names from the first entry in result
	fieldNames := make([]string, 0, len(result[0]))
	for key := range result[0] {
		fieldNames = append(fieldNames, key)
	}

	for _, fieldName := range fieldNames {
		// Extract values for each field across all result entries
		var fieldValues interface{}

		switch result[0][fieldName].(type) {
		case int, int8, int16, int32, int64, float32, float64:
			values := make([]float64, len(result))
			for i, entry := range result {
				values[i] = entry[fieldName].(float64)
			}
			fieldValues = values

		case string:
			values := make([]string, len(result))
			for i, entry := range result {
				// Sanitize specific field values if needed
				if fieldName == "new" || fieldName == "value:display" {
					values[i] = SanitizeValues([]string{fmt.Sprintf("%v", entry[fieldName])})[0]
				} else {
					values[i] = entry[fieldName].(string)
				}
			}
			fieldValues = values

		case time.Time:
			values := make([]time.Time, len(result))
			for i, entry := range result {
				values[i] = entry[fieldName].(time.Time)
			}
			fieldValues = values

		default:
			// If type is unknown, default to string
			values := make([]string, len(result))
			for i, entry := range result {
				values[i] = fmt.Sprintf("%v", entry[fieldName])
			}
			fieldValues = values
		}

		// Create a new field and add it to the frame
		frame.Fields = append(frame.Fields, data.NewField(fieldName, nil, fieldValues).SetConfig(&data.FieldConfig{DisplayName: fieldName}))
	}

	if DebugLevel() == 1 {
		PrintDebug("You are Inside mapTextResponseToFrame")
		PrintDebug(frame)
	}

	return frame
}
